docs(cache-handlers): drop entries on expire, not revalidate (#96517)
vercel/next.js: docs/01-app/03-api-reference/05-config/01-next-config-js/cacheHandlers.mdx

Reviewer: Not sure if it's even relevant, but would it be possible to add usage of `durations` to this example? Or to another example, if it's for example only relevant for an in-memory cache handler?

Author: Added — the distributed example's markers are now `{ stale, expired }` pairs, mirroring what the built-in handler stores: no `durations` (e.g. `updateTag()`) → hard-expire now; with `durations` (profile-based, e.g. `revalidateTag(tag, 'max')`) → stale now, hard-expire `durations.expire` seconds later (`Infinity` → never). `getExpiration` reports only the `expired` stamp, since that's the one the runtime discards on — the `stale` stamp is a `get()`-side concern, where the built-in handler serves the entry but returns it with `revalidate: -1` to force a background refresh.

Reviewer: Same here, would it be possible to add example usage of `softTags` in whatever example it would make sense to use it?

Author: Good call — added it to this example: `get()` now checks the entry's own tags plus the incoming `softTags` against `tag:*` stamps and drops the entry when any tag is newer. To make that coherent I also switched this example's `getExpiration` to the documented `Infinity` sentinel ("soft tags are checked in `get`") and made `updateTags` actually write the stamps — the runtime special-cases `Infinity` and keeps implicit tags unexpired on its side, so the `get`-side check is the single enforcement point in this variant.

Reviewer: When would one use this "Distributed Tag Coordination" setup? For the Redis example, with tracking invalidations in Redis and checking of `softTags` in `get()`, my understanding is that Redis is the single source of truth regardless of multiple Next.js instances?

Author: Right — in both patterns Redis stays the single source of truth. The mirror isn't a second truth, it's a per-request cache of it, and the difference is purely where reads pay.

Checking tags inside `get()` (the external-storage example) costs an extra Redis round-trip on every cache read. Perfectly correct, simplest to write, and at moderate traffic it's all you need. But a single request can hit dozens of `'use cache'` reads, plus `getExpiration()` for the implicit tags — with in-`get()` checks that's dozens of extra round-trips per request. The mirror flips that: `refreshTags()` syncs tag state once per request (that's the reason this method exists in the interface at all), and every read in that request checks tags at memory speed. What you give up is a propagation window of at most one request boundary for invalidations issued by other instances.

So: start with in-`get()` checks, reach for the mirror when per-read latency or Redis load starts to matter. Added a paragraph to the section intro saying exactly this, since I think every reader will land on your question sooner or later.

Reviewer: Cool, that makes sense, thanks!

Returns a `CacheEntry` object if found, or `undefined` if not found or expired.

Your `get` method should retrieve the cache entry from storage, check if it has expired based on the `expire` time, and return `undefined` for missing or expired entries.

> Return the entry even when it is past its `revalidate` time. `revalidate` marks the entry as stale, not expired: Next.js serves the stale entry and schedules a background refresh. Only `expire` means the entry can no longer be used.

Reviewer 2: ```suggestion
Your `get` method should retrieve the cache entry from storage and return `undefined` only for missing entries or entries past their `expire` time. Return entries that are only past `revalidate`, since those are stale rather than expired, and returning them lets Next.js serve them while refreshing in the background.
```

Author: Applied, thanks — reads better as one paragraph.

```js
const cacheHandler = {
  async get(cacheKey, softTags) {
    const entry = cache.get(cacheKey)
    if (!entry) return undefined

    // Only `expire` means the entry is unusable. An entry past its
    // `revalidate` time is stale, not expired: return it so Next.js can

>     // serve it while refreshing in the background.

Reviewer 2: ```suggestion
   // Check if expired
```

Author: Applied.

      return undefined
    }

    // Only drop the entry once it is past `expire`. Past `revalidate` it is

>     // stale, and returning it lets Next.js serve it while it refreshes.

Reviewer 2: ```suggestion
    // Check if entry has expired
```

Author: Applied.

        expire: entry.expire,
        revalidate: entry.revalidate,
      }),
      // Use Redis TTL for automatic expiration. `EX` requires a positive
      // integer, while `expire` can be fractional or `Infinity` (the

>       // never-expire sentinel), so clamp before passing it through.

Reviewer 2: ```suggestion
      // `EX` needs a positive integer.
      // By definition `expire: Infinity` ("never expire")
      // maps to no TTL at all, so omit `EX` and let Redis persist the key.
```

Author: Applied. Kept the `Math.ceil` in the code itself since `expire` can also be fractional, comment as suggested.

        localTagTimestamps.set(tagKeys[i], JSON.parse(values[i]))
      }
    }
  },

  async getExpiration(tags) {
    // Report the hard-expiration stamps; Next.js discards entries older
    // than the returned value. Stale stamps are consumed in `get()`
    // instead: serve the entry, but return it with `revalidate: -1` to
    // force a background refresh (this mirrors the built-in handler).
    const stamps = tags.map((tag) => localTagTimestamps.get(tag)?.expired ?? 0)
    return Math.max(...stamps, 0)
  },

  async updateTags(tags, durations) {
    const now = Date.now()
    // Without `durations` (for example `updateTag()`), the tag expires
    // immediately: entries carrying it must not be served again. With
    // `durations` (profile-based, for example `revalidateTag(tag, 'max')`),
    // entries turn stale now — keep serving them while they refresh in the
    // background — and only hard-expire `durations.expire` seconds later.
    const expired =
      durations === undefined
        ? now
        : Number.isFinite(durations.expire)
          ? now + durations.expire * 1000
          : 0 // `expire: Infinity` — never hard-expire

>     const marker = JSON.stringify({ stale: now, expired })

Reviewer 2: This whole thing is scope creeping without a holistic view I think, for example now, the `stale` key, as it is, is not used right? We'd need to expand line 474... and compare timestamp to stale and expire? 

Right now the only other get method that could possible use this, is actually dropping them no? 

I know having a more concrete example is good, but having way too specific examples, might be better for an example, and even that one can go so far only. 

Wydt?

Reviewer 2: @leejpsd anything work to say here? — Kind of want to merge this and start trying to get the example ready to merge too

Author: Sorry for the slow reply — this sat far longer than it should have.

You're right, and I'd rather cut it than defend it. The `stale` stamp isn't read anywhere: that example's `get()` is elided, so anyone following it would store a field nothing consumes. It also left the two examples disagreeing — the external-storage `get()` drops on a newer stamp while this one implied serve-stale, without ever showing it.

So the distributed example is back to plain timestamps, and the `durations` semantics moved up into the `updateTags()` reference, where it's two bullets of contract instead of a half-built state machine: absent means invalidate now (`updateTag()`, single-arg `revalidateTag()`), present means stale now and still servable until `expire` passes (`revalidateTag(tag, 'max')`). That's what @Svish was asking for and it belongs there rather than in an example anyway.

Net effect is the page is smaller than my previous revision. Ready to merge from my side, and happy to take another pass if anything still reads as too much.

Author: One more small push after that: dropped the now-unused `durations` parameter from the distributed example's `updateTags`, so both concrete examples take just what they use. Left the signature as-is in the API reference example, since showing the full shape belongs there.

That's everything from me.
